### PR TITLE
VehicleHead::updateLoadCargo

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Ultimately very similar to the unload function. Not done much testing of this yet so may have some bugs.

### CODE REVIEW

**Review by AaronVanGeffen — CHANGES_REQUESTED** [review]
Diverges compared to master.

First month:
```
Hex Compare
Produced: 10/06/21 10:57:52 am 

Left file: /home/aaron/.config/OpenLoco/save/autosave/base 2050/autosave_2021-06-10_10-56-34.SV5
Right file: /home/aaron/.config/OpenLoco/save/autosave/cargoload/autosave_2021-06-10_10-55-02.SV5
2247051 same byte(s)
7465 left orphan byte(s)
7846 right orphan byte(s)
19186 difference byte(s)
```

Second month:
```
Hex Compare
Produced: 10/06/21 10:58:12 am 

Left file: /home/aaron/.config/OpenLoco/save/autosave/base 2050/autosave_2021-06-10_10-56-55.SV5
Right file: /home/aaron/.config/OpenLoco/save/autosave/cargoload/autosave_2021-06-10_10-55-21.SV5
2169911 same byte(s)
29334 left orphan byte(s)
28731 right orphan byte(s)
62551 difference byte(s)
```

Third month:
```
Hex Compare
Produced: 10/06/21 10:58:23 am 

Left file: /home/aaron/.config/OpenLoco/save/autosave/base 2050/autosave_2021-06-10_10-57-14.SV5
Right file: /home/aaron/.config/OpenLoco/save/autosave/cargoload/autosave_2021-06-10_10-55-36.SV5
2129338 same byte(s)
31239 left orphan byte(s)
29408 right orphan byte(s)
72222 difference byte(s)
```

**Comment by duncanspumpkin** [comment]
Could you recheck i think ive caught the sources of divergence.

**Review by AaronVanGeffen — CHANGES_REQUESTED** [review]
Less divergence, but not gone.

First month:
```
Hex Compare
Produced: 21/06/21 7:30:29 pm 

Left file: /home/aaron/.config/OpenLoco/save/autosave/base 2050/autosave_2021-06-10_10-56-34.SV5
Right file: /home/aaron/.config/OpenLoco/save/autosave/cargoload/autosave_2021-06-21_19-28-11.SV5
2273588 same byte(s)
11 left orphan byte(s)
4 right orphan byte(s)
103 difference byte(s)
```

Second month:
```
Hex Compare
Produced: 21/06/21 7:30:21 pm 

Left file: /home/aaron/.config/OpenLoco/save/autosave/base 2050/autosave_2021-06-10_10-56-55.SV5
Right file: /home/aaron/.config/OpenLoco/save/autosave/cargoload/autosave_2021-06-21_19-28-31.SV5
2245175 same byte(s)
5688 left orphan byte(s)
5143 right orphan byte(s)
10933 difference byte(s)
```


Third month:
```
Hex Compare
Produced: 21/06/21 7:29:31 pm 

Left file: /home/aaron/.config/OpenLoco/save/autosave/base 2050/autosave_2021-06-10_10-57-14.SV5
Right file: /home/aaron/.config/OpenLoco/save/autosave/cargoload/autosave_2021-06-21_19-28-49.SV5
2206512 same byte(s)
9544 left orphan byte(s)
7797 right orphan byte(s)
16743 difference byte(s)
```

**Comment by duncanspumpkin** [comment]
some sort of issue with company expenditures but not sure if thats the route cause.

**Comment by duncanspumpkin** [comment]
Diverges on station cargo stats rating and quantity as well.

File: src/OpenLoco/Vehicles/VehicleHead.cpp [hunk 1]
```diff
@@ -2806,26 +2827,272 @@ namespace OpenLoco::Vehicles
...
+        station->updateCargoDistribution();
+        cargo.qty += qtyTransferred;
+        const uint8_t typeAgeMap[] = { 0, 5, 3, 2, 0, 0 };
+        stationCargo.age = std::min(stationCargo.age, typeAgeMap[static_cast<uint8_t>(vehicleType)]);
+
+        station->var_3B0 = 0;
+        station->flags |= StationFlags::flag_7;
+        Vehicle train(this);
+        auto vehMaxSpeed = train.veh2->maxSpeed;
+        auto carAgeFactor = std::max<uint32_t>(0xFF, (getCurrentDay() - bogie->creation_day) / 256);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Vehicles/VehicleHead.cpp:2974]: min!

**Comment by AaronVanGeffen** [comment]
Fwiw:

First month:
```
Hex Compare
Produced: 22/06/21 9:32:06 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050/autosave_2021-06-10_10-56-34.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload/autosave_2021-06-22_21-29-43.SV5
2273682 same byte(s)
3 left orphan byte(s)
4 right orphan byte(s)
17 difference byte(s)
```

Second month:
```
Hex Compare
Produced: 22/06/21 9:31:47 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050/autosave_2021-06-10_10-56-55.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload/autosave_2021-06-22_21-30-04.SV5
2261792 same byte(s)
4 difference byte(s)
```

Third month:
```
Hex Compare
Produced: 22/06/21 9:31:02 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050/autosave_2021-06-10_10-57-14.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload/autosave_2021-06-22_21-30-23.SV5
2227203 same byte(s)
1848 left orphan byte(s)
1755 right orphan byte(s)
3748 difference byte(s)
```

**Comment by duncanspumpkin** [comment]
I'm not getting divergence anymore when playing the lets play save for 4 months (except for differences that can be accounted for with the frame rate)

**Comment by AaronVanGeffen** [comment]
Comparing master to this PR, with #936 cherry-picked in to uncompress saves:

First month:
```
Hex Compare
Produced: 23/06/21 10:42:37 am 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050 uncompressed/autosave_2021-06-23_10-41-03.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload uncompressed/autosave_2021-06-23_10-38-00.SV5
6783948 same byte(s)
13 difference byte(s)
```

Second month:
```
Hex Compare
Produced: 23/06/21 10:42:59 am 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050 uncompressed/autosave_2021-06-23_10-41-23.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload uncompressed/autosave_2021-06-23_10-38-20.SV5
6785364 same byte(s)
5 difference byte(s)
```

Third month:
```
Hex Compare
Produced: 23/06/21 10:43:17 am 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050 uncompressed/autosave_2021-06-23_10-41-41.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload uncompressed/autosave_2021-06-23_10-38-38.SV5
6776858 same byte(s)
4973 left orphan byte(s)
4973 right orphan byte(s)
3650 difference byte(s)
```

Fourth month:
```
Hex Compare
Produced: 23/06/21 10:44:05 am 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050 uncompressed/autosave_2021-06-23_10-41-58.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload uncompressed/autosave_2021-06-23_10-38-58.SV5
6587918 same byte(s)
195198 left orphan byte(s)
195198 right orphan byte(s)
3029 difference byte(s)
```

**Comment by AaronVanGeffen** [comment]
Looks like you've curbed the divergence. Comparing master to this PR, with #936 cherry-picked in to uncompress saves:

First month:
```
Hex Compare
Produced: 24/06/21 1:59:17 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050 uncompressed/autosave_2021-06-23_10-41-03.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload uncompressed/autosave_2021-06-24_13-56-33.SV5
6783943 same byte(s)
18 difference byte(s)
```

Second month:
```
Hex Compare
Produced: 24/06/21 1:59:34 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050 uncompressed/autosave_2021-06-23_10-41-23.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload uncompressed/autosave_2021-06-24_13-56-53.SV5
6785364 same byte(s)
5 difference byte(s)
```

Third month:
```
Hex Compare
Produced: 24/06/21 1:59:45 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050 uncompressed/autosave_2021-06-23_10-41-41.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload uncompressed/autosave_2021-06-24_13-57-12.SV5
6785472 same byte(s)
9 difference byte(s)
```

Fourth month:
```
Hex Compare
Produced: 24/06/21 1:59:58 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/base 2050 uncompressed/autosave_2021-06-23_10-41-58.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/cargoload uncompressed/autosave_2021-06-24_13-57-30.SV5
6786140 same byte(s)
5 difference byte(s)
```

**Comment by AaronVanGeffen** [comment]
So much for it being 'dead' code. Comparing master to this PR, with #936 cherry-picked in to uncompress saves:

First month:
```
Hex Compare
Produced: 25/06/21 11:30:26 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/matsku-2050-cargoload-uncompressed-2/autosave_2021-06-25_23-26-38.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/matsku-2050-master-uncompressed/autosave_2021-06-24_22-35-43.SV5
6783943 same byte(s)
18 difference byte(s)
```

Second month:
```
Hex Compare
Produced: 25/06/21 11:30:35 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/matsku-2050-cargoload-uncompressed-2/autosave_2021-06-25_23-26-59.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/matsku-2050-master-uncompressed/autosave_2021-06-24_22-36-04.SV5
6785365 same byte(s)
4 difference byte(s)
```

Third month:
```
Hex Compare
Produced: 25/06/21 11:30:43 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/matsku-2050-cargoload-uncompressed-2/autosave_2021-06-25_23-27-19.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/matsku-2050-master-uncompressed/autosave_2021-06-24_22-36-21.SV5
6785472 same byte(s)
9 difference byte(s)
```

Fourth month:
```
Hex Compare
Produced: 25/06/21 11:30:54 pm 

Left file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/matsku-2050-cargoload-uncompressed-2/autosave_2021-06-25_23-27-38.SV5
Right file: /home/aaron/Repositories/OpenLoco/OpenLoco/build/config/save/autosave/matsku-2050-master-uncompressed/autosave_2021-06-24_22-36-39.SV5
6786141 same byte(s)
4 difference byte(s)
```

**Comment by AaronVanGeffen** [comment]
On g++ builds, there's a divergence in game behaviour on 28618af (second commit), but this goes away as soon as the dead code is introduced in 7a3e817. We verified that this code is not actually executed, however.

On clang++ builds, both commits are fine, and do not diverge from master. I'm thinking this could be due to some kind of compiler optimisation, but I'm not sure how best to pursue this. Any ideas?

**Comment by AaronVanGeffen** [comment]
I have retested this PR, compiling three versions using GCC 11.1.0:
1. master (06e6c34671c953b24158ce72cd408c187ae124db) + cherry-picked uncompress saves commit (5a337ddb2bcd2592fd8bf06e631b6006db51bdae)
2. cargoload HEAD (5a337ddb2bcd2592fd8bf06e631b6006db51bdae) 
3. cargoload HEAD (5a337ddb2bcd2592fd8bf06e631b6006db51bdae) + revert dead code commit (9e4619ab070f628b9fb3df4d28a1221011e50535)

For each of these builds, I generated four autosaves by letting the Matsku 2050 save game run for four in game months (matsku-2050-master-uncompressed).

Comparing these autosaves, I find _no divergence_, in contrast to what was reported earlier. I'm a bit puzzled why, but I checked everything twice.

I propose we drop the last two commits (dead code and uncompress saves), then merge this PR.

**Comment by duncanspumpkin** [comment]
Rebased and dropped the 2 commits.

**Review by AaronVanGeffen — APPROVED** [review]

**Pull request merged**